base_layer/core/src/base_node/states/fetching_horizon_state.rs

```diff
@@ -49,60 +81,185 @@ impl HorizonInfo {
+        let mut base_state = MutableMmrLeafNodes::new(Vec::new(), Bitmap::create());
+        while index < complete_leaf_count {
+            let MutableMmrState {
+                total_leaf_count,
+                leaf_nodes,
+            } = shared
+                .comms
+                .fetch_mmr_state(tree.clone(), index, leaf_nodes_sync_chunk_size)
+                .await
+                .map_err(|e| e.to_string())?;
+            base_state.combine(leaf_nodes);
+            complete_leaf_count = total_leaf_count as u64;
+            index += leaf_nodes_sync_chunk_size;
+        }
+        Ok(base_state)
+    }
+
+    async fn synchronize_headers<B: BlockchainBackend>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+    ) -> Result<(), String>
+    {
+        let height_indices = (0..self.horizon_block).collect::<Vec<u64>>();
+        for block_nums in height_indices.chunks(shared.config.horizon_sync_config.headers_sync_chunk_size) {
+            let headers = shared
+                .comms
+                .fetch_headers(block_nums.to_vec())
+                .await
+                .map_err(|e| e.to_string())?;
+
+            let mut txn = DbTransaction::new();
+            headers.into_iter().for_each(|header| txn.insert_header(header));
+            shared.db.commit(txn).map_err(|e| e.to_string())?;
+        }
+        Ok(())
     }
 
-    async fn synchronize_kernels(&mut self) -> Result<(), String> {
-        Err("unimplemented".into())
+    async fn synchronize_kernels<B: BlockchainBackend>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+    ) -> Result<(), String>
+    {
+        let mmr_state = shared
+            .db
+            .fetch_mmr_base_leaf_nodes(MmrTree::Kernel, 0, 1)
+            .map_err(|e| e.to_string())?;
+        let kernel_hashes = shared
+            .db
+            .fetch_mmr_base_leaf_nodes(MmrTree::Kernel, 0, mmr_state.total_leaf_count)
+            .map_err(|e| e.to_string())?
+            .leaf_nodes
+            .leaf_hashes;
+
+        for hashes in kernel_hashes.chunks(shared.config.horizon_sync_config.kernels_sync_chunk_size) {
+            let kernels = shared
+                .comms
+                .fetch_kernels(hashes.to_vec())
+                .await
+                .map_err(|e| e.to_string())?;
+
+            let mut txn = DbTransaction::new();
+            kernels.into_iter().for_each(|kernel| txn.insert_kernel(kernel));
+            shared.db.commit(txn).map_err(|e| e.to_string())?;
+        }
+        Ok(())
     }
 
-    async fn synchronize_utxo_set(&mut self) -> Result<(), String> {
-        Err("unimplemented".into())
+    async fn synchronize_utxo_set<B: BlockchainBackend>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+    ) -> Result<(), String>
+    {
+        let mmr_state = shared
+            .db
+            .fetch_mmr_base_leaf_nodes(MmrTree::Utxo, 0, 1)
+            .map_err(|e| e.to_string())?;
```

Comment:
Is the purpose of this to get the leaf node count?

If so, we should definitely add a DB query to return this and then tidy this function up.

Alternatively, `count` could become an `Option<usize>` and if it's `None` we default to returning all the leaf nodes.

Reply:
Yes, the purpose was to get the leaf node count. I added the `fetch_mmr_base_leaf_node_count` function to the backend to clean this code.